### PR TITLE
Deprecate TLS_MAX_VERSION and DTLS_MAX_VERSION

### PR DESCRIPTION
**Pull request by kroeckx**
Fixes: #7183

### CODE REVIEW

**Comment by t-j-h** [comment]
Is the usage noted in the issue such that having a function returning this (as suggested) is not needed?
It seems to be to be half of a fix for the issue ...

**Comment by kroeckx** [comment]
I could find 2 users for it:
- It called SSL_CTX_set_max_proto_version() with TLS_MAX_VERSION. That code had been changed to use 0 instead.
- It used some complicated code that should get replaced with a call to SSL_CTX_set_min_proto_version() when build against 1.2.0 because the rest is deprecated (but only in the manual).

**Comment by mattcaswell** [comment]
Doesn't this mean we'll start getting deprecation warnings for our own libssl code as soon as we hit 1.2.0?

**Comment by kroeckx** [comment]
The TLS_MAX_VERSION define would move to an internal header when we remove it from the public header.

**Comment by mattcaswell** [comment]
But it can't be in both at the same time. While its in the public header inside deprecation guards, won't we get deprecation warnings from the compiler - which will cause a build failure with --strict-warnings?

**Comment by kroeckx** [comment]
You mean at the moment we change the version?
Does that mean I should rename it for internal use, say MAX_TLS_VERSION.

**Comment by mattcaswell** [comment]
> You mean at the moment we change the version?

Yes

> Does that mean I should rename it for internal use, say MAX_TLS_VERSION.

Probably - yes.

**Comment by kroeckx** [comment]
It really doesn't make sense to me to have a different name for the SSL_OP_NO_* to use internally, yet have the same value.

**Comment by tmshort** [comment]
MAX_TLS_VERSION vs. TLS_MAX_VERSION is bound to cause confusion...

**Comment by mattcaswell** [comment]
`TLS_MAX_VERSION_INTERNAL`?

**Comment by kroeckx** [comment]
I've rebased this and added a new commit

**Comment by kroeckx** [comment]
So it now fails in the no-deprecated build because things like SSL_OP_NO_TLSv1_3 are used internally. As long as they're deprecated they have to be used internally to be able to still support them.

**Comment by kroeckx** [comment]
Does someone have some comment on this? The only thing I see is to have an other copy of those defines, which I don't like.

**Comment by mattcaswell** [comment]
> So it now fails in the no-deprecated build because things like SSL_OP_NO_TLSv1_3 are used internally. As long as they're deprecated they have to be used internally to be able to still support them.

Looking at the internal usage, it all seems fairly straight forward to remove, rewrite or put inside `OPENSSL_API_3` guards, with the exception of the stuff in ssl_conf. The "Protocol" conf command and the "-no_tls1_3" (and similar) command line switches are essentially thin wrappers around these options. If we deprecate the config options then probably we also have to deprecate the conf command and command line switches. Obviously we *could* do it with internal definitions instead - but since SSL_CONF still gives the user the ability to set these things in another way it kind of defeats the object of deprecation to do so.

My guess is the "Protocol" command and the "-no_\*" switches are fairly widely known and used - so deprecating them might be a step too far. If so then that implies deprecating `SSL_OP_NO_*` might also be a step too far.

I'd be interested in @vdukhovni's opinion.

**Comment by vdukhovni** [comment]
We must not remove the `SSL_OP_NO_...` protocol de-selection options too quickly. They're ugly, but they've been the sole protocol selection interface for a long time.

Furthermore, applications will generally assume that when the option is gone, it means that the protocol in question is gone, and no longer needs to be disabled, but deprecation will mean that the protocol will remain enabled, and the application should have used the cleaner min/max versions (which I instigated way back, so I'd like to see them used, but...).

So my concern is that removing the macros will lead to insecure configurations, rather than compile-time failure.  So we need to be very careful when removing these.  Postfix, for example, is not yet ready for min/max, and still presents the protocol-by-protocol de-selection interface to users.  I'll be addressing that for upcoming releases, but it will take a while (~4-5 years) before we drop support for older code that only knows the `SSL_OP_NO_...` approach.

File: include/openssl/ssl.h [hunk 1]
```diff
@@ -374,18 +374,21 @@ typedef int (*SSL_verify_cb)(int preverify_ok, X509_STORE_CTX *x509_ctx);
...
-# define SSL_OP_NO_TLSv1_1                               0x10000000U
-# define SSL_OP_NO_TLSv1_3                               0x20000000U
-
-# define SSL_OP_NO_DTLSv1                                0x04000000U
-# define SSL_OP_NO_DTLSv1_2                              0x08000000U
-
-# define SSL_OP_NO_SSL_MASK (SSL_OP_NO_SSLv3|\
-        SSL_OP_NO_TLSv1|SSL_OP_NO_TLSv1_1|SSL_OP_NO_TLSv1_2|SSL_OP_NO_TLSv1_3)
-# define SSL_OP_NO_DTLS_MASK (SSL_OP_NO_DTLSv1|SSL_OP_NO_DTLSv1_2)
+# if !OPENSSL_API_3
```
Comment (vdukhovni) [inline include/openssl/ssl.h]: So bottom line, I think that `API_3` is perhaps too soon to lose the `SSL_OP_NO_...` protocol disablement bits.

**Comment by kroeckx** [comment]
I would like to make clear that they are deprecated other than
just the documentation saying so. I agree that we don't want to
remove them soon. I guess I'll just drop that commit for now.

**Comment by kroeckx** [comment]
So I've pushed a new version

**Pull request closed without merging**